File: src/main/java/work/lclpnet/ap2/api/base/GameQueue.java
package work.lclpnet.ap2.api.base;

import work.lclpnet.ap2.api.game.MiniGame;

import java.util.List;

public interface GameQueue {

    MiniGame pollNextGame();

    List<MiniGame> preview();

    void setNextGame(MiniGame miniGame);

    void shiftGame(MiniGame miniGame);
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/impl/base/GameQueue.kt
package work.lclpnet.ap2.impl.base

import work.lclpnet.ap2.game.MiniGame

interface GameQueue {
    fun pollNextGame(): MiniGame

    fun preview(): List<Entry>

    fun setNextGame(miniGame: MiniGame)

    fun shiftGame(miniGame: MiniGame)

    fun setFilter(filter: (MiniGame) -> Boolean)

    fun updateHistory(game: MiniGame)

    enum class Type {
        REGULAR,
        VOTED,
        PRIORITY
    }

    data class Entry(
        val game: MiniGame,
        val type: Type,
    )
}
